pinc/special_colors.inc

```diff
@@ -1,63 +1,91 @@
 <?php
 include_once($relPath.'misc.inc');
 
+// return a set of hard-coded special days keyed by spec_code
+function load_common_special_days()
+{
+    return [
+        "_BIRTHDAY_RECENT" => [
+            'color' => "CCFFFF",
+            'display_name' => "Authors with recent birthdays",
+        ],
+        "_BIRTHDAY_TODAY" => [
+            'color' => "33CCFF",
+            'display_name' => "Authors with birthdays today",
+        ],
+        "_OTHER_SPECIAL" => [
+            'color' => "FFFF66",
+            'display_name' => "Other Special",
+        ],
+    ];
+}
+
+function load_special_days()
+{
+    $specials_array = load_common_special_days();
+
+    $sql = "
+        SELECT spec_code, color, display_name
+        FROM special_days
+        ORDER BY display_name
+    ";
+    $specs_result = DPDatabase::query($sql);
+
+    while ($row = mysqli_fetch_assoc($specs_result)) {
+        $specials_array[$row['spec_code']] = [
+            "color" => $row['color'],
+            "display_name" => $row['display_name'],
+        ];
+    }
+
+    return $specials_array;
+}
 
 // Returns the special color associated with this
 // project, or null if no such color is specified.
 //
 // $book is supposed to be an associative array
 // representing a  record from the
 // projects table. At the moment it is 'enough'
-// that the keys 'comments' and 'username' are
-// existant.
-function get_special_color_for_project($book) {
-
-    $bgcolor = null;
-
+// that the key 'special_code' exists.
+function get_special_color_for_project($book)
+{
     $special_code = $book['special_code'];
 
-    // first three are (for now) special cases, the rest will be data-driven
-    // from the SPECIAL_DAYS table
+    static $specials_array = [];
```

Comment:
would it make sense to set this outside the function, that way we could avoid setting up a similar variable on line 109?

Reply:
I don't want to make it a global variable (global variables = evil).

I would have liked to pull the `load_common_special_days()` array directly into `load_special_days()` and be done with it, but the logic in `get_special_color_for_project()` requires we keep them separate since we can't rely solely on the project's `special_code` value to determine which special day it goes with. ☹️